### PR TITLE
Replicate Methodhandle.invoke/invokeExact behavior in OJDK impl

### PR DESCRIPTION
**Pull request by fengxue-IS**
Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

**Comment by fengxue-IS** [comment]
FYI @babsingh , local testing in progress

File: runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp [hunk 1]
```diff
@@ -1522,5 +1522,18 @@ Java_java_lang_invoke_MethodHandleNatives_registerNatives(JNIEnv *env, jclass cl
 	return;
 }
 
+jobject JNICALL
+Java_java_lang_invoke_MethodHandle_invokeExact(JNIEnv *env, jclass ignored, jobject handle, jobject args)
```
Comment (DanHeidinga) [inline runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:1526]: Isn't this already present in the java_dyn_methodhandle.c file:
https://github.com/eclipse/openj9/blob/7fade16235106505d2d5a52a940d0fa0511ce76c/runtime/jcl/common/java_dyn_methodhandle.c#L792

Comment (DanHeidinga) [reply runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:1526]: Can the two be commoned?

**Comment by fengxue-IS** [comment]
@DanHeidinga can you please take another look.
I've merged the JCL headers and I wasn't able to think of a good location to common the methods, do you have any suggestions?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by babsingh** [comment]
@fengxue-IS The conflicts in this PR will need to be resolved.

> I wasn't able to think of a good location to common the methods

I do not think there is a good location where common code between OJ9 and OJDK MHs can be placed. You can create a new source file named `java_dyn_methodhandle_common.cpp`, which will compile these methods for both implementations. 

For compilation, you will need to update https://github.com/eclipse-openj9/openj9/blob/master/runtime/jcl/CMakeLists.txt#L170.

For exporting symbols, you will need to create `java_dyn_methodhandle_common_exports.xml` similar to `java_dyn_methodhandle_exports.xml`.

This will allow you to share (common) the code between the two implementation.

**Comment by fengxue-IS** [comment]
fixed merge conflict.

@babsingh given that these two methods only throw an UnsupportedOperationException, I am not sure if we should dedicate a new file + buildfile to it.

**Comment by babsingh** [comment]
Semantically, `java_lang_invoke_MethodHandleNatives.cpp` should only include native methods from the `MethodHandleNatives` class.

**Another option:** instead of not compiling `java_dyn_methodhandle.c` for OJDK MHs, we can `ifdef` the source code and exports, and compile it with just the common/shared methods.

File: runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp (hunk 1 above)
Comment (tajila) [reply runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:1600]: Is it possible to build both all the time, and rely on the ifdefs for exclusion between ojdkmh and j9mh?

File: runtime/jcl/uma/java_lang_invoke_MethodHandleNatives_exports.xml [hunk 2]
```diff
@@ -33,4 +33,6 @@
 	<export name="Java_java_lang_invoke_MethodHandleNatives_clearCallSiteContext" />
 	<export name="Java_java_lang_invoke_MethodHandleNatives_getNamedCon" />
 	<export name="Java_java_lang_invoke_MethodHandleNatives_registerNatives" />
+	<export name="Java_java_lang_invoke_MethodHandle_invoke" />
```
Comment (tajila) [inline runtime/jcl/uma/java_lang_invoke_MethodHandleNatives_exports.xml:36]: Likewise, if we build both we can put the exports in se829_exports.xml

File: runtime/jcl/exports.cmake [hunk 3]
```diff
@@ -629,6 +629,8 @@ if(J9VM_OPT_OPENJDK_METHODHANDLE)
 		Java_java_lang_invoke_MethodHandleNatives_clearCallSiteContext
 		Java_java_lang_invoke_MethodHandleNatives_getNamedCon
 		Java_java_lang_invoke_MethodHandleNatives_registerNatives
+		Java_java_lang_invoke_MethodHandle_invoke
```
Comment (tajila) [inline runtime/jcl/exports.cmake:632]: you can do a similar change here (like with UMA) and just define these in the first export block.

Also, you can remove the exports in line 442,443

**Comment by tajila** [comment]
> @babsingh given that these two methods only throw an UnsupportedOperationException, I am not sure if we should dedicate a new file + buildfile to it.

Instead of adding a new file, one of the files can be built all the time and guarded with `#ifdefs`. I propose the new MH file as the older one will eventually be obsolete

**Comment by fengxue-IS** [comment]
> Instead of adding a new file, one of the files can be built all the time and guarded with `#ifdefs`. I propose the new MH file as the older one will eventually be obsolete

I will try setting up `java_lang_invoke_MethodHandleNatives.cpp` to use ifdef to compile both versions

**Comment by tajila** [comment]
@fengxue-IS Need to update coprights on java_dyn_methodhandle

```
15:51:12  The following files were modified and have incorrect copyrights
[Pipeline] echo
15:51:12  runtime/jcl/common/java_dyn_methodhandle.c
```

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
jenkins test sanity winojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by tajila** [comment]
jenkins test sanity xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by fengxue-IS** [comment]
@tajila all testing passed, can this be merged?

**Pull request merged**